File: 프로그래머스/2/12936. 줄 서는 방법/줄 서는 방법.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(int n, long long k) {
    vector<int> nums;
    vector<int> answer;
    long long count = 1;
    for(int i = 1; i <= n; ++i)
    {
        nums.push_back(i);
        count *= i;
    }
    count /= n;
    --n;
    --k;
    while(true)
    {
        long long target = (k / count);
        answer.push_back(nums[target]);
        nums.erase(nums.begin() + target);
        k = k % count;
        
        if (n < 1)
            break;
        
        count /= n--; 
    }
    
    return answer;
}

File: 프로그래머스/2/12936. 줄 서는 방법/줄 서는 방법.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {

    public static int[] solution(int n, long k) {
        int[] answer = new int[n];
        List<Integer> nums = new ArrayList<>();

        for (int i = 1; i <= n; i++) {
            nums.add(i);
        }

        long num = 1;

        for(int i = 1; i < n; ++i)
        {
            num *= i;
        }
        
        --n;
        --k;
        int targetIdx = 0;
        while(!nums.isEmpty())
        {
            int idx = (int)(k / num);
            answer[targetIdx] = nums.get(idx);
            nums.remove(idx);
            k %= num;
            
            if (n == 0)
                break;
            
            num /= n;
            --n;
            ++targetIdx;
        }
        return answer;
    }

}
